core/trino-main/src/main/java/io/trino/operator/DriverFactory.java

```diff
@@ -93,16 +93,20 @@ public List<OperatorFactory> getOperatorFactories()
         return operatorFactories;
     }
 
-    public synchronized Driver createDriver(DriverContext driverContext)
+    public Driver createDriver(DriverContext driverContext)
     {
-        checkState(!noMoreDrivers, "noMoreDrivers is already set");
         requireNonNull(driverContext, "driverContext is null");
-        List<Operator> operators = new ArrayList<>();
+        List<Operator> operators = new ArrayList<>(operatorFactories.size());
         try {
-            for (OperatorFactory operatorFactory : operatorFactories) {
-                Operator operator = operatorFactory.createOperator(driverContext);
-                operators.add(operator);
+            synchronized (this) {
+                // must check noMoreDrivers after acquiring the lock
+                checkState(!noMoreDrivers, "noMoreDrivers is already set");
+                for (OperatorFactory operatorFactory : operatorFactories) {
+                    Operator operator = operatorFactory.createOperator(driverContext);
+                    operators.add(operator);
+                }
             }
+            // Driver creation can continue without holding the lock
```

Comment:
Could you please explain why this change is needed? Is this just an efficiency performance? Or was there a deadlock?

Reply:
It didn't create deadlocks, but it more tightly integrates the synchronized section to what's actually being guarded, which is synchronizing callers to `operatorFactories.createOperator()` and `operatorFactories.noMoreOperators()` (on each). The `checkState(!noMoreDrivers, ...)` call is always going to be prone to races between checking `DriverFactory#isNoMoreDrivers()` and `DriverFactory#createDriver()`.

So this _specific_ piece is just to constrain the locked region to the minimum part, but the other changes to make `isNoMoreDrivers()` not synchronize and `DriverSplitRunnerFactory#createDriver` handle the exception that could result from a race-prone check against `noMoreDrivers` instead of failing is more important.